Partly remove support for seccomp annotations

saschagrunert opened the pull request:
#### What type of PR is this?

/kind api-change

#### What this PR does / why we need it:

We now partly drop the support for seccomp annotations which is planned for v1.25 as part of the KEP:

Making the annotations fully non-functional will be deferred to a future release.

#### Which issue(s) this PR fixes:

Refers to https://github.com/kubernetes/enhancements/issues/135
Docs PR: https://github.com/kubernetes/website/pull/33524
Fixes https://github.com/kubernetes/kubernetes/issues/95171

#### Special notes for your reviewer:

Pod security policies are not touched by this change and therefore we have to keep the annotation key constants.

Warnings emitted for annotation-only use since 1.22 in https://github.com/kubernetes/kubernetes/pull/102491, and specific warnings about removal of support in 1.25 emitted since 1.23 in https://github.com/kubernetes/kubernetes/pull/104389

#### Does this PR introduce a user-facing change?

```release-note
Action required: support for the alpha seccomp annotations `seccomp.security.alpha.kubernetes.io/pod` and `container.seccomp.security.alpha.kubernetes.io`, deprecated since v1.19, has been partially removed. Kubelets no longer support the annotations, use of the annotations in static pods is no longer supported, and the seccomp annotations are no longer auto-populated when pods with seccomp fields are created. Auto-population of the seccomp fields from the annotations is planned to be removed in 1.27. Pods should use the corresponding pod or container `securityContext.seccompProfile` field instead.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
KEP: https://github.com/kubernetes/enhancements/issues/135
```

saschagrunert committed:
Partly remove support for seccomp annotations

We now partly drop the support for seccomp annotations which is planned
for v1.25 as part of the KEP:

https://github.com/kubernetes/enhancements/issues/135

Pod security policies are not touched by this change and therefore we
have to keep the annotation key constants.

This means we only allow the usage of the annotations for backwards
compatibility reasons while the synchronization of the field to
annotation is no longer supported. Using the annotations for static pods
is also not supported any more.

Making the annotations fully non-functional will be deferred to a
future release.

Signed-off-by: Sascha Grunert <sgrunert@redhat.com>

saschagrunert commented:
/test pull-kubernetes-node-e2e-containerd

liggitt commented on pkg/api/pod/warnings.go:
```diff
@@ -234,21 +234,7 @@ func warningsForPodSpecAndMeta(fieldPath *field.Path, podSpec *api.PodSpec, meta
 		warnings = append(warnings, fmt.Sprintf("%s: fractional byte value %q is invalid, must be an integer", fieldPath.Child("spec", "overhead").Key(string(api.ResourceEphemeralStorage)), value.String()))
 	}
 
-	// use of pod seccomp annotation without accompanying field
```
we should keep emitting these warnings to make it clear they're doing something that isn't working as they expect

liggitt commented on pkg/api/pod/warnings.go:
```diff
@@ -234,21 +234,7 @@ func warningsForPodSpecAndMeta(fieldPath *field.Path, podSpec *api.PodSpec, meta
-	}
-
 	pods.VisitContainersWithPath(podSpec, fieldPath.Child("spec"), func(c *api.Container, p *field.Path) bool {
-		// use of container seccomp annotation without accompanying field
```
same here, keep emitting these

liggitt commented on pkg/registry/core/pod/strategy.go:
```diff
@@ -646,74 +643,3 @@ func validateContainer(container string, pod *api.Pod) (string, error) {
-
-// applySeccompVersionSkew implements the version skew behavior described in:
-// https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/135-seccomp#version-skew-strategy
-func applySeccompVersionSkew(pod *api.Pod) {
```
@tallclair do you have thoughts on the best way to phase this out?

the two audiences I'm considering are:
* manifest authors / API clients
* security scanners / policy enforcers / admission webhooks

ways we could phase out annotation support (unordered):
* stop copying annotations to fields
* stop copying fields to annotations
* stop requiring fields and annotations to match
* reject things that set annotations and not fields
* stop validating seccomp annotations completely

a quick sweep of manifests using the alpha annotations shows that most consumers in github have not backfilled their manifests to include seccomp fields (including manifests in kubernetes and kubernetes-sigs orgs)

I'd be curious to know whether policy enforcers are inspecting both annotations and fields, or just annotations, or just fields. if just annotations, what would the consequences be of letting those enforcers think (based on an annotation) that seccomp was in use when in fact (based on the absence of a seccomp field or a non-matching field) that seccomp policy was not in use?

tallclair replied:
@saschagrunert what's the status of seccomp-by-default? From https://github.com/kubernetes/enhancements/pull/3240 it looks like we're planning to bring it to beta in v1.25, which I assume means default-enabled? If so, I think that mitigates a lot of jordan's concerns about non-updated manifests, since I expect the vast majority of those are setting the default profile.

Here's some numbers from searching public repos:
- `runtime/default`: ~1.8K matches https://sourcegraph.com/search?q=context:global+seccomp%5C.security%5C.alpha%5C.kubernetes%5C.io.*runtime/default+count:2000&patternType=regexp
- `docker/default`: ~5.1K matches https://sourcegraph.com/search?q=context:global+seccomp%5C.security%5C.alpha%5C.kubernetes%5C.io.*docker/default+count:10000&patternType=regexp
- `unconfined`: 663 matches https://sourcegraph.com/search?q=context:global+seccomp%5C.security%5C.alpha%5C.kubernetes%5C.io.*unconfined+count:10000&patternType=regexp
- `localhost`: 258 matches https://sourcegraph.com/search?q=context:global+seccomp%5C.security%5C.alpha%5C.kubernetes%5C.io.*localhost/+count:10000+-repo:kubernetes/kubernetes&patternType=regexp

IMO as long as this is paired with seccomp-by-default, an action required release note should be sufficient to roll this out. One other thing we could consider is updating the [baseline pod-security check](https://github.com/kubernetes/kubernetes/blob/cc6619857027c634918a6d2cca9bab072c46f71f/staging/src/k8s.io/pod-security-admission/policy/check_seccompProfile_baseline.go#L29-L48) to forbid the use of any seccomp annotations.

saschagrunert replied:
> @saschagrunert what's the status of seccomp-by-default? From [kubernetes/enhancements#3240](https://github.com/kubernetes/enhancements/pull/3240) it looks like we're planning to bring it to beta in v1.25, which I assume means default-enabled?

Since Kubernetes v1.24 beta features are not enabled by default any more. I'll reflect that in the KEP, but still aim to graduate the feature to beta in v1.25 to make it API aware.

tallclair replied:
Ah, I missed that it's opt-in. That's too bad.

In that case, I think for v1.25 we should:
- Stop copying fields to annotations on create
- Warn when setting the annotation

And then push this change out a couple releases, at which point I'd be OK with either ignoring annotations (stop copying them to the field) or explicitly forbidding them.

tallclair replied:
Also the warning should probably be applied to workload resources too if we want anyone to see it.

Another thing we could do: update the pod security restricted profile to forbid setting the seccomp annotations in v1.25+

liggitt replied:
> I think for v1.25 we should:
>
> * Stop copying fields to annotations on create
> * Warn when setting the annotation
>
> And then push this change out a couple releases, at which point I'd be OK with either ignoring annotations (stop copying them to the field) or explicitly forbidding them.

That seems like the right direction to wean people off annotations.

I'd probably keep the validation that says if field and annotation are both set, they have to match, but gradually remove the auto-copying between them (remove field → annotation copying in 1.25, remove annotation → field copying in 1.X)

> Also the warning should probably be applied to workload resources too if we want anyone to see it.

any warning we put in GetWarningsForPod will surface for built-in workloads containing pod specs

> Another thing we could do: update the pod security restricted profile to forbid setting the seccomp annotations in v1.25+

Mmm... that feels sort of weird. I'd consider keeping the validation that says if you specify both, they have to match.

tallclair replied:
> remove annotation → field copying in 1.X

I don't think this is worth doing. I don't want anyone looking at annotations, so I think at the point we stop copying annotations to fields we should consider annotations done, and remove the annotation code entirely.

I agree that we should keep the validation that says if both are set they must match.

liggitt replied:
>> remove annotation → field copying in 1.X
>
> I don't think this is worth doing

Did I misread this:

> And then push this change out a couple releases, at which point I'd be OK with either ignoring annotations (stop copying them to the field)...

I thought that's what you were suggesting

tallclair replied:
Ah, yeah, I think I meant ignoring them holistically, basically what this PR is currently doing.

saschagrunert replied:
So I guess we're fine here, right?

tallclair replied:
My suggestion was:

> In that case, I think for v1.25 we should:
> - Stop copying fields to annotations on create
> - Warn when setting the annotation

And then push the changes in this PR (completely ignoring annotations) out a release or more.

saschagrunert replied:
> My suggestion was:
> 
> > In that case, I think for v1.25 we should:
> > 
> > * Stop copying fields to annotations on create
> > * Warn when setting the annotation
> 
> And then push the changes in this PR (completely ignoring annotations) out a release or more.

@liggitt are you fine with that?

liggitt replied:
> In that case, I think for v1.25 we should:
>
> 1. Stop copying fields to annotations on create
> 2. Warn when setting the annotation

That sounds good for 1.25.

> And then push the changes in this PR (completely ignoring annotations) out a release or more.

I'm not sure we should ever drop the validation that forbids a mismatch between annotation and field if both are set. I think it's reasonable to drop the annotation → field copying after a bit.

saschagrunert commented:
/test pull-kubernetes-e2e-kind

saschagrunert commented:
/test pull-kubernetes-e2e-kind

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -169,7 +169,6 @@ func ValidatePodSpecificAnnotations(annotations map[string]string, spec *core.Po
 		}
 	}
 
-	allErrs = append(allErrs, ValidateSeccompPodAnnotations(annotations, fldPath)...)
```
this looks right to me, would like a second set of eyes from @tallclair

saschagrunert replied:
Added  the test cases for now.

saschagrunert commented on pkg/api/pod/util.go:
```diff
@@ -467,6 +471,13 @@ func GetValidationOptionsFromPodSpecAndMeta(podSpec, oldPodSpec *api.PodSpec, po
 	}
 
+	if oldPodMeta != nil && oldPodSpec != nil {
+		if len(validation.ValidateSeccompPodAnnotations(oldPodMeta.Annotations, field.NewPath("metadata").Child("annotations"))) > 0 &&
+			len(validation.ValidateSeccompAnnotationsAndFields(*oldPodMeta, oldPodSpec, field.NewPath("template").Child("spec"))) > 0 {
```
Yes this makes sense :+1:

saschagrunert commented:
/test pull-kubernetes-e2e-kind

endocrimes commented:
/triage accepted

saschagrunert commented:
@tallclair please take a look at the open discussion points :)

cbandy commented on staging/src/k8s.io/api/core/v1/annotation_key_constants.go:
```diff
@@ -36,12 +36,12 @@ const (
 	// SeccompContainerAnnotationKeyPrefix represents the key of a seccomp profile applied
 	// to one container of a pod.
-	// Deprecated: set a container security context `seccompProfile` field.
+	// Deprecated: set a container security context `seccompProfile` field. The key is non functional in 1.25+.
```
Is this comment still accurate? It looks like some code does read the annotations.

https://github.com/kubernetes/kubernetes/pull/109819/files#r880206105

> I added a patch … for copying the annotations to the fields (and not the other way around).

saschagrunert replied:
No you're right. I removed the additional comment because I think the exact behavioral description should be part of the official docs.

cbandy commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -19793,10 +19794,46 @@ func TestValidatePodTemplateSpecSeccomp(t *testing.T) {
 			},
 		},
+		{
+			description: "seccomp field and container annotation must not match when invalid seccomp annotations are allowed",
```
Is "must" accurate here? Should this say "may not match" when invalid allowed?

saschagrunert replied:
Sure, fixed as suggested :+1:

saschagrunert commented:
/test pull-kubernetes-e2e-kind-ipv6

saschagrunert commented:
@liggitt @tallclair please take another look

saschagrunert commented:
@kubernetes/api-approvers PTAL

tallclair commented on pkg/registry/core/pod/strategy.go:
```diff
@@ -644,28 +644,20 @@ func validateContainer(container string, pod *api.Pod) (string, error) {
 	// get possible annotation and field
 	annotation, hasAnnotation := pod.Annotations[v1.SeccompPodAnnotationKey]
-	field, hasField := (*api.SeccompProfile)(nil), false
+	_, hasField := (*api.SeccompProfile)(nil), false
```
```suggestion
	hasField := false
```

tallclair replied:
Alternatively:
```suggestion
	hasField := (pod.Spec.SecurityContext != nil && pod.Spec.SecurityContext.SeccompProfile != nil)
```

saschagrunert replied:
Going for simplicity and applying the first suggestion. Thanks for that!

tallclair commented on pkg/registry/core/pod/strategy.go:
```diff
@@ -682,24 +674,14 @@ func applySeccompVersionSkew(pod *api.Pod) {
 			annotation, hasAnnotation := pod.Annotations[key]
 
-			field, hasField := (*api.SeccompProfile)(nil), false
+			_, hasField := (*api.SeccompProfile)(nil), false
```
Same as above
```suggestion
			hasField := false
```
or 
```suggestion
			hasField := (ctr.SecurityContext != nil && ctr.SecurityContext.SeccompProfile != nil)
```

tallclair reviewed (commented):
Can you update the release note to be more specific with what this PR is changing?

Sorry for the delinquent review!

saschagrunert commented on pkg/api/pod/util.go:
```diff
@@ -436,6 +437,8 @@ func GetValidationOptionsFromPodSpecAndMeta(podSpec, oldPodSpec *api.PodSpec, po
 		AllowOSField: utilfeature.DefaultFeatureGate.Enabled(features.IdentifyPodOS),
 		// The default sysctl value does not contain a forward slash, and in 1.24 we intend to relax this to be true by default
 		AllowSysctlRegexContainSlash: false,
+		// Do not allow the usage of invalid seccomp annotations per default. It will be set to true only when ValidateSeccompPodAnnotations and ValidateSeccompAnnotationsAndFields fails.
```
Fixed as suggested.

saschagrunert commented on pkg/api/pod/warnings.go:
```diff
@@ -237,15 +237,15 @@ func warningsForPodSpecAndMeta(fieldPath *field.Path, podSpec *api.PodSpec, meta
 	// use of pod seccomp annotation without accompanying field
 	if podSpec.SecurityContext == nil || podSpec.SecurityContext.SeccompProfile == nil {
```
> I think we should always return the warning when the annotation is set, otherwise users will never see it since we copy the annotation to the field. Same applies below (per-container warning).

Makes sense, I'll remove the checks for `podSpec.SecurityContext == nil || podSpec.SecurityContext.SeccompProfile == nil`

> Also, is this called on PodTemplates embedded in workload resources? I think it would be good to return a warning in those cases too, as that's the most common way to create pods.

Yes, see: https://github.com/kubernetes/kubernetes/blob/2a017f94bcf8d04cbbbbdc6695bcf74273d630ed/pkg/api/pod/warnings.go#L60

saschagrunert commented on pkg/api/pod/warnings.go:
```diff
@@ -237,15 +237,15 @@ func warningsForPodSpecAndMeta(fieldPath *field.Path, podSpec *api.PodSpec, meta
 	// use of pod seccomp annotation without accompanying field
 	if podSpec.SecurityContext == nil || podSpec.SecurityContext.SeccompProfile == nil {
 		if _, exists := meta.Annotations[api.SeccompPodAnnotationKey]; exists {
-			warnings = append(warnings, fmt.Sprintf(`%s: deprecated since v1.19, non-functional in v1.25+; use the "seccompProfile" field instead`, fieldPath.Child("metadata", "annotations").Key(api.SeccompPodAnnotationKey)))
```
Sounds good, fixed.

saschagrunert commented:
@tallclair thank you for the review! I updated the release notes and incorporated the changes.

tallclair commented:
/lgtm
/approve

saschagrunert commented:
Thank you!

Please take a look for:
- API approval: @deads2k @lavalamp @smarterclayton @wojtek-t @liggitt
- kubelet approval: @Random-Liu @dchen1107 @derekwaynecarr @yujuhong @sjenning @mrunalp @klueska

saschagrunert commented:
Had to rebase.

saschagrunert commented:
/test pull-kubernetes-e2e-kind

saschagrunert commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

saschagrunert commented:
> Can we capture in the commit message what we agreed to in the conversations above?

Yep, rebased and changed the commit message to contain the additional context.

saschagrunert commented:
/test pull-kubernetes-e2e-kind

saschagrunert commented:
/test pull-kubernetes-e2e-kind

saschagrunert commented:
/test pull-kubernetes-e2e-kind

liggitt commented on pkg/api/pod/warnings.go:
```diff
@@ -234,19 +234,15 @@ func warningsForPodSpecAndMeta(fieldPath *field.Path, podSpec *api.PodSpec, meta
 		warnings = append(warnings, fmt.Sprintf("%s: fractional byte value %q is invalid, must be an integer", fieldPath.Child("spec", "overhead").Key(string(api.ResourceEphemeralStorage)), value.String()))
 	}
 
-	// use of pod seccomp annotation without accompanying field
```
I would limit this change to adjusting the warning message.

If they're setting both the annotation and the field, I still don't think we need to warn them.

liggitt commented on pkg/api/pod/warnings.go:
```diff
@@ -234,19 +234,15 @@ func warningsForPodSpecAndMeta(fieldPath *field.Path, podSpec *api.PodSpec, meta
 
 	pods.VisitContainersWithPath(podSpec, fieldPath.Child("spec"), func(c *api.Container, p *field.Path) bool {
-		// use of container seccomp annotation without accompanying field
-		if c.SecurityContext == nil || c.SecurityContext.SeccompProfile == nil {
```
same comment here about not needing to warn if they're setting both the annotation and the field

liggitt commented:
The more I think about it, the less value I see in *ever* dropping validation of the annotations if they're present, or requiring annotations and fields to match if they're both present (https://github.com/kubernetes/kubernetes/pull/109819/files#r879450837). As users stop setting the annotations, that code will go dormant, but as long as they are setting annotations (possibly in manifests that target both new and very old clusters), making sure the data is coherent seems good.

I think this PR should limit itself to:
* dropping kubelet support for annotations in `pkg/kubelet/...`
* dropping copying fields → annotations in `applySeccompVersionSkew`
* dropping the pod-level annotation from `isSignificantPodUpdate`
* tweak the "annotations aren't effective in 1.25+" language in `warningsForPodSpecAndMeta`

In a future release, we would drop `applySeccompVersionSkew` entirely, but leave the annotation validation and field<->annotation validation as-is.

saschagrunert commented:
Thank you for the review @liggitt, I changed `pkg/api/pod/warnings.go` as suggested and reverted the related test changes.

saschagrunert commented:
/test pull-kubernetes-integration

liggitt commented on pkg/api/pod/util.go:
```diff
@@ -766,36 +777,6 @@ func SeccompAnnotationForField(field *api.SeccompProfile) string {
 
-// SeccompFieldForAnnotation takes a pod annotation and returns the converted
-// seccomp profile field.
-func SeccompFieldForAnnotation(annotation string) *api.SeccompProfile {
```
relocating this function looks good

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3413,6 +3415,8 @@ type PodValidationOptions struct {
 	// Allow more DNSSearchPaths and longer DNSSearchListChars
 	AllowExpandedDNSConfig bool
+	// Allow the invalid (but non-functional) seccomp annotations for backwards compatibility support.
+	AllowInvalidSeccompAnnotations bool
```
I think we should drop this change and not relax validation in the future. The annotation validation will simply go unused as people stop specifying the annotations, but anyone who continues to set them will continue to have the same validation applied. (Dropping this makes this PR a lot smaller as well)

saschagrunert replied:
Alright I dropped this part and reverted the tests.

cbandy commented:
> The more I think about it, the less value I see in _ever_ dropping validation of the annotations if they're present, or requiring annotations and fields to match if they're both present (https://github.com/kubernetes/kubernetes/pull/109819/files#r879450837). As users stop setting the annotations, that code will go dormant, but as long as they are setting annotations (possibly in manifests that target both new and very old clusters), making sure the data is coherent seems good.
> 
> I think this PR should limit itself to:
> 
> * dropping kubelet support for annotations in `pkg/kubelet/...`
> * dropping copying fields → annotations in `applySeccompVersionSkew`
> * dropping the pod-level annotation from `isSignificantPodUpdate`
> * tweak the "annotations aren't effective in 1.25+" language in `warningsForPodSpecAndMeta`


I've gotten a little lost in all this, so let me restate my understanding of the timeline for seccomp annotations:
- 1.3: Kubelet respects annotations on the container (and sandbox?)
- 1.19, GA: Validate and sync annotations to and from fields.
- 1.22: Warn when annotations are used without fields.
- 1.23: Warn that annotations will cease to function as of 1.25.
- 1.25, this PR: Stop copying fields to annotations.

If I understand correctly, the intent at this point is that in 1.25:
- When present, annotations are copied into missing fields with a warning.
- When present, annotations are validated to match fields.
- Kubelet respects [only] fields on the container (and sandbox?)

🤔 This seems like a smaller change than I expected based on the language we used to have in the warning. I suspect that is because we should have done more work in the Kubelet circa 1.22, when the annotation-field sync was everywhere.

Looking forward, where Kubelet respects only fields in release N: (N = 1.25 if this PR goes well.)
- N+2: Stop copying annotations into missing fields.
- N+2: Stop considering annotations in Pod Security Admission.

> In a future release, we would drop `applySeccompVersionSkew` entirely, but leave the annotation validation and field<->annotation validation as-is.

Would this be on or after 1.27, which is N+2 in my thinking above?

My sense is that we don't have to leave the annotation code _forever_, but I could be wrong. Perhaps a few versions later (N+4?) we could:
- Stop validating that annotations match fields.
- Stop warning when annotations are used without fields.

The [KEP to GA the fields](https://git.k8s.io/enhancements/keps/sig-node/135-seccomp) details how to deprecate the annotations but does not describe the steps necessary to remove them (after deprecation). Does it warrant another KEP? Are there other examples to follow?

liggitt commented:
/lgtm
/approve

liggitt commented:
> I've gotten a little lost in all this, so let me restate my understanding of the timeline for seccomp annotations:
> 
> * 1.3: Kubelet respects annotations on the container (and sandbox?)
> * 1.19, GA: Validate and sync annotations to and from fields.
> * 1.22: Warn when annotations are used without fields.
> * 1.23: Warn that annotations will cease to function as of 1.25.
> * 1.25, this PR: Stop copying fields to annotations.
> 
> If I understand correctly, the intent at this point is that in 1.25:
> 
> * When present, annotations are copied into missing fields with a warning.
> * When present, annotations are validated to match fields.
> * Kubelet respects [only] fields on the container (and sandbox?)

All of that looks correct, expect the last bit - Kubelet respects the fields on the pod and container, just not the annotations.

> > In a future release, we would drop `applySeccompVersionSkew` entirely, but leave the annotation validation and field<->annotation validation as-is.
> 
> Would this be on or after 1.27, which is N+2 in my thinking above?

Yes, likely 1.27.

> My sense is that we don't have to leave the annotation code _forever_, but I could be wrong. Perhaps a few versions later (N+4?) we could:
> 
> * Stop validating that annotations match fields.
> * Stop warning when annotations are used without fields.

I just don't see much benefit to dropping that code and allowing in data that some API clients could be confused/broken by.
